File: Baekjoon/greedy/241029_3085. 사탕게임.java
import java.io.BufferedReader;
import java.io.InputStreamReader;


public class Main {

    static int N, ans;
    static char[][] map;

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        N = Integer.parseInt(br.readLine());
        map = new char[N][N];

        for (int i=0; i<N; i++) {
            String inputStr = br.readLine();  // 띄어쓰기로 구분된 것이 아니라면 String 타입으로 입력을 받아 chatAt()으로 접근한다
            for (int j=0; j<N; j++) {
                map[i][j] = inputStr.charAt(j);
            }
        }

        findAns();

        System.out.println(ans);
    }

    // 그리디 사용: 바꾸었을 때, 바꾸지 않았을 때의 결과 비교
    static void findAns() {
        for (int i=0; i<N; i++) {
            for (int j=0; j<N; j++) {
                if (j+1 < N) {
                    swap(i, j, i, j+1);
                    checkMaxLength();
                    swap(i, j, i, j+1);  // 복구한다
                }

                if (i+1 < N) {
                    swap(i, j, i+1, j);
                    checkMaxLength();
                    swap(i, j, i+1, j);  // 복구한다
                }
            }
        }
    }

    static void swap(int x1, int y1, int x2, int y2) {
        char temp = map[x1][y1];
        map[x1][y1] = map[x2][y2];
        map[x2][y2] = temp;
    }

    static void checkMaxLength() {
        // 가로 방향에서 찾기
        for (int i=0; i<N; i++) {
            int rowMax = 1;
            for (int j=0; j<N-1; j++) {
                if (map[i][j] == map[i][j+1]) {
                    rowMax++;
                } else {
                    ans = Math.max(ans, rowMax);
                    rowMax = 1;
                }
            }
            ans = Math.max(ans, rowMax);
        }

        // 세로 방향에서 찾기
        for (int i=0; i<N; i++) {
            int colMax = 1;
            for (int j=0; j<N-1; j++) {
                if (map[j][i] == map[j+1][i]) {
                    colMax++;
                } else {
                    ans = Math.max(ans, colMax);
                    colMax = 1;
                }
            }
            ans = Math.max(ans, colMax);
        }
    }
}

File: javascript/241029_3085. 사탕게임.js
const readline = require("readline");

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

let N;
let map = [];
let ans = 0;

rl.on("line", (line) => {
  if (!N) {
    N = parseInt(line); // 첫 번째 입력은 N
  } else {
    map.push(line.split("")); // 이후 입력은 보드 상태를 저장
    if (map.length === N) {
      // 모든 입력을 받으면 계산 시작
      findAns();
      console.log(ans);
      rl.close();
    }
  }
});

function findAns() {
  for (let i = 0; i < N; i++) {
    for (let j = 0; j < N; j++) {
      // 오른쪽과 교환
      if (j + 1 < N) {
        swap(i, j, i, j + 1);
        checkMaxLength();
        swap(i, j, i, j + 1); // 원상복구
      }

      // 아래쪽과 교환
      if (i + 1 < N) {
        swap(i, j, i + 1, j);
        checkMaxLength();
        swap(i, j, i + 1, j); // 원상복구
      }
    }
  }
}

function swap(x1, y1, x2, y2) {
  const temp = map[x1][y1];
  map[x1][y1] = map[x2][y2];
  map[x2][y2] = temp;
}

function checkMaxLength() {
  // 가로 방향에서 연속된 사탕 찾기
  for (let i = 0; i < N; i++) {
    let rowMax = 1;
    for (let j = 0; j < N - 1; j++) {
      if (map[i][j] === map[i][j + 1]) {
        rowMax++;
      } else {
        ans = Math.max(ans, rowMax);
        rowMax = 1;
      }
    }
    ans = Math.max(ans, rowMax);
  }

  // 세로 방향에서 연속된 사탕 찾기
  for (let j = 0; j < N; j++) {
    let colMax = 1;
    for (let i = 0; i < N - 1; i++) {
      if (map[i][j] === map[i + 1][j]) {
        colMax++;
      } else {
        ans = Math.max(ans, colMax);
        colMax = 1;
      }
    }
    ans = Math.max(ans, colMax);
  }
}
